platforms/core-runtime/stdlib-java-extensions/src/main/java/org/gradle/internal/function/Predicate.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.internal.function;
+
+/**
+ * Represents some predicate against objects of type T.
+ *
+ * @param <T> The target type for this predicate
+ */
+public interface Predicate<T> {
```

Comment:
💭  Replacement for `Spec<T>` that does not require depending on core-api

Reply:
>Replacement for Spec<T> that does not require depending on core-api

🤔 Have you considered moving `Spec` to this module instead?

This would be at least fourth abstraction of this kind available (`java.util.Predicate`, guava's `Predicate`, and the `Spec` itself being the other three). Or is the plan to replace it with `java.util` later, when we drop Java 7 support?

Reply:
> 🤔 Have you considered moving Spec to this module instead?

It crossed my mind briefly, but I decided against it. Mainly because `stdlib-java-extensions` currently contains next to no types that are user facing. Upon further inspection prompted by your question, I was actually surprised to find `JavaVersion` in there. Unless we decide to move more API types there, I wouldn't do it.

Another argument is that the explicit usage of the new `Predicate` is very limited: only Java 6 modules that use `CollectionUtils`. One of them is `GUtil`, which will go away in Gradle 9.0; probably way before we drop support for Java 6.

Given that there are no other major problems, I'd proceed with the merge, and then check in with the team about moving `Spec` into `stdlib-java-extensions`. It'll be a quick refactor, if we decide in favor.

Reply:
Hm, the `JavaVersion` type seems to be in `org.gradle.api` only historically, and may not be intended for use by Gradle users either. This makes `stdlib-java-extentions` internal "by intent". So I'd make a separate decision about starting to move intended public types in there.